{"cells":[{"cell_type":"markdown","source":["Nom Etudiant 1:\n","<br>\n","Nom Etudiant 2:"],"metadata":{"id":"e6SSCja7My_c"}},{"cell_type":"markdown","metadata":{"id":"Rx466dVIpn3_"},"source":["# TME 2.2 : Median String\n","\n","\n","## Partie A : Recheche de pattern (motifs) en permettant des variations\n","\n","Les motifs que nous cherchons dans les sequences d'ADN peuvent contenir quelques variations ou mutations. Nous allons developpé dans cet partie differents algoritmes pour la recherche de motifs variables. "]},{"cell_type":"markdown","metadata":{"id":"fdcJxBtVpn4F"},"source":["1\\. Comme dans le TME precedent, nous alons d'abord générer des données atificielles pour pouvoir tester les algorithmes. Faire une fonction pour générer `t` séquences artificielles de taille `n`. Implantez dans chaque séquence un motif de taille `k` à des positions aléatoires avec `v` substitutions choisies aléatoirement."]},{"cell_type":"code","execution_count":null,"metadata":{"id":"LSz7T4a3pn4G"},"outputs":[],"source":["import random\n","\n","nuc = ('A', 'C', 'G', 'T')\n","\n","k=4 #taille de motif\n","v=1 #nb de positions variable dans le motif\n","t=3 #nb de sequences\n","n=10 #longuer des sequence\n","\n","def insertMotif(sequence, motif, position):\n","    return sequence[:position] + motif + sequence[position:]\n","\n","def generateRandomSequence(n:int, upper=False):\n","    \"\"\"\n","    Génère une séquence nucléotidique aléatoire \n","    entrée n : longueur de la sequence\n","    entrée upper : bool, si True les nucléotides seront minuscule, False majuscule\n","    sortie sequence : une séquence nucléotidique aléatoire \n","    \"\"\"\n","    sequence = \"\"\n","    \n","    return sequence\n","\n","def modifierMotif(motif:str, nbpos:int,  upper=False):\n","    \"\"\"\n","    Modifie nbpos positions d'un motif aléatoirement \n","    entrée motif: motif à modifier\n","    entrée nbpos: nombre de positions\n","    entrée upper : bool, si True les nucléotides modifiés seront minuscule, False majuscule\n","    sortie motifM: motif modifié\n","    \"\"\"\n","    motifM = list(motif)\n","    \n","    return \"\".join(motifM)\n","\n","#tester modifMotif\n","#print (modifMotif(\"acg\", 2))\n","\n","\n","def implantMotifVar(k, v, t, n):\n","    \"\"\"\n","    Génère des séquences aléatoires et les implante des motifs variables (un motif par séquence)\n","    entrée k: taille du motif\n","    entrée v: nombre de variations\n","    entrée t : nombre de séquences \n","    entrée n : longueur des séquences\n","    sortie DNA : matrice de dimension txn avec les motifs implantés\n","    REMARQUE : La taille totale des séquences plus motif doit être égal à t, pensez à générer de séquence aléatoire de taille t-k pour pouvoir implanter un motif de taille k\n","    \"\"\"\n","    \n","    sequences = []\n","    \n","    return sequences\n","\n","adn = implantMotifVar(k, v, t, n)\n","print (adn)\n"]},{"cell_type":"markdown","metadata":{"id":"G0AARLzmpn4I"},"source":["2\\. Visualisation de motifs. Nous pouvons visualiser les motifs à l'aide des outils de LOGO https://weblogo.berkeley.edu/logo.cgi. Executer votre fonction de generation de motif variable, extraire les motifs et visulise-le à l'aide de webLogo. Utiliser les parametre ci-dessous.\n"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"TYCLBSw8pn4J"},"outputs":[],"source":["k=8 #taille de motif\n","v=2 #nb de positions variable dans le motif\n","t=10 #nb de sequences\n","n=10 #longuer des sequence\n","\n","adn = implantMotifVar(k, v, t, n)\n","print (adn)"]},{"cell_type":"markdown","metadata":{"id":"LhfgfAkipn4Q"},"source":["3\\. Implémentez l'algorithme _\"Median String Search\"_ pour chercher des motifs de taille variable. Tester le sur les jeux de données gnerer precedement. Vous devez éliminer les motifs peu complexe pour éviter les calculs inutiles. Pensez à eliminer tous les motifs ayant une formation diferents de motifs attendues, par exemple, les motifs ayant deux base repeté comme ACACACAC, etc.  "]},{"cell_type":"code","execution_count":null,"metadata":{"id":"oq0K6y6rpn4Q"},"outputs":[],"source":["from itertools import product\n","\n","\n","def removeLowComplexe(motifs):\n","   \n","    return validMotif\n","\n","\n","\n","\n","                \n","#Genere tous les K-mers de taille K ayant de AAA... à TTT...\n","allkmers = product(nuc, repeat=k)\n","kmers = kmerList(allkmers)\n","\n","\n","print (len(kmers))\n","kmersValid = removeLowComplexe(kmers, 4)\n","print (len(kmersValid))\n","print (kmersValid[0])\n","\n"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"e5biEcQGpn4R"},"outputs":[],"source":["def hamdist(str1:str, str2:str):\n","\t\"\"\"\n","\tCalcul la distance de hamming entre deux chaînes de caractères\n","\tentrée str1: chaîne de caractères\n","\tentrée str2: chaîne de caractères\n","\tsortie distance: distance de hamming\n","\t\"\"\"\n","\tdiffs = 0\n","\t\n","\treturn diffs\n","\n","\n","def totalDistance(motif:str, sequences, k):\n","\t\"\"\"\n","\tCalcul la totalDistance\n","\tentrée motif: sequence consensus\n","\tentrée sequences: chaîne de caractères\n","\tentrée k: taille du motif \n","\tentrée sequences: matrice de dimension txn contenant les séquences\n","\tsortie totalDistance: somme de distance de hamming minimal\n","\t\"\"\"\n","\ttd = 0\n","\t\n","\treturn td\n","\n","\n","def MedianStringSearch(allkmers, sequences, k):\n","\t\"\"\"\n","\tImplement l'algorithme MedianStringSearch\n","\tentrée allkmers: liste de K-mers valides\n","\tentrée sequences: matrice de dimension txn contenant les séquences\n","\tentrée k: taille du motif \n","\tsortie bestMotif: le motif que minimise les distances\n","\tsortie bestDistance: la distance minimal\n","\tsortie motifDist: un dictionnaire contenant les motifs et leurs distances\n","\t\"\"\"\n","\tbestDistance = 1000\n","\tmotifDist = {}\n","\tcount = 0\n","\tfor kmer in allkmers:\n","\t\n","\treturn bestMotif, bestDistance, motifDist\n","\n"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"8cV411yOpn4R"},"outputs":[],"source":["#Apliquez l'algoritme MedianStringSearch sur les sequences du brin complementaire\n","\n","def reverseComplement(seq):\n","    seq_dict = {'A':'T','T':'A','G':'C','C':'G'}\n","    return \"\".join([seq_dict[base] for base in reversed(seq)])\n","\n","\n"]},{"cell_type":"markdown","source":["## Partie B : Recherche de motifs variable sur vos données"],"metadata":{"id":"tgUYBJ0yQ3Dg"}},{"cell_type":"markdown","source":["Chercher sur le fichier \"Sequence_by_Peaks_G*.fasta\" contiens les regions de peak trouvé par ChipSeq, qui contient probablement le Facteur de Transcription que nous cherchons. Apliquer l'alogorithm Median Sting pour chercher les motifs dans votre fichier \"Sequence_by_Peaks_G*. Il faut bien evidement enlever les motifs peu complexe."],"metadata":{"id":"I1zHn6NSQ2Xy"}},{"cell_type":"code","source":["import time\n","\n","k=7\n","feq=10\n","top= 50\n","genome = \"Sequence_by_Peaks_1.fasta\"\n","\n","\n","\n","def readFasta(fastaFileName):\n","    \"\"\"\n","    Read a fasta file\n","    entrée fastaFileName: nom du fichier fasta\n","    sortie sequences: liste contenant toutes les sequences du fichier\n","    \"\"\"\n","    sequence = []\n","    file = open(genome, \"r\")\n","    sequence = []\n","    for s in file:\n","        if s[0] != \">\":\n","            sequence.append(s.strip().upper())\n","\n","    return sequence\n","\n","sequences   = readFasta(genome)\n"],"metadata":{"id":"A6KINeGuREqA"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["bestMotif, bestDistance, motifDist = MedianStringSearch(kmersValid, sequences[0:3], k)\n","print (bestMotif, bestDistance)"],"metadata":{"id":"Y7W7nNWvRn_R"},"execution_count":null,"outputs":[]}],"metadata":{"colab":{"provenance":[]},"kernelspec":{"display_name":"Python 3","language":"python","name":"python3"},"language_info":{"codemirror_mode":{"name":"ipython","version":3},"file_extension":".py","mimetype":"text/x-python","name":"python","nbconvert_exporter":"python","pygments_lexer":"ipython3","version":"3.7.4"}},"nbformat":4,"nbformat_minor":0}